### ISSUE TITLE
out of my depth on this one

### ISSUE THREAD

**Issue by unicomp21** [murat-dogan/node-datachannel#201]
root@c1e4da2a2e0f:~/repo/dev/monorepo/root/blazor/nodedatachannel# npx tsx server.ts 
node:internal/errors:478
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /root/repo/dev/monorepo/root/blazor/nodedatachannel/node_modules/node-datachannel/package.json
    at __node_internal_captureLargerStackTrace (node:internal/errors:478:5)
    at new NodeError (node:internal/errors:387:5)
    at throwExportsNotFound (node:internal/modules/esm/resolve:365:9)
    at packageExportsResolve (node:internal/modules/esm/resolve:589:7)
    at resolveExports (node:internal/modules/cjs/loader:556:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:596:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1014:27)
    at Function.l.default._resolveFilename (/root/repo/dev/monorepo/root/blazor/nodedatachannel/node_modules/tsx/dist/cjs/index.cjs:1:1671)
    at Function.Module._load (node:internal/modules/cjs/loader:873:27)
    at Module.require (node:internal/modules/cjs/loader:1100:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at <anonymous> (/root/repo/dev/monorepo/root/blazor/nodedatachannel/server.ts:4:45)
    at Object.<anonymous> (/root/repo/dev/monorepo/root/blazor/nodedatachannel/server.ts:78:2)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.j (/root/repo/dev/monorepo/root/blazor/nodedatachannel/node_modules/tsx/dist/cjs/index.cjs:1:1197)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:169:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:530:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}

**Comment by unicomp21** [comment]
trying to run under tsx

**Comment by dguenther** [comment]
I also encountered this when updating to v0.5.0. node-datachannel has switched to ES Modules: https://github.com/murat-dogan/node-datachannel/blob/53551d14f87a475dd900acc1ea53a9146c2530aa/package.json#L5 

As a result, to import node-datachannel from a CommonJS-based package, you'll have to use dynamic imports, as documented here: https://nodejs.org/api/esm.html#import-expressions

**Comment by joecarl** [comment]
I solved it importing it as:
```js
import * as RTC from 'node-datachannel';
``` 
And then using it as:
```js
const conn = new RTC.PeerConnection('pc', ...);
```

**Comment by dguenther** [comment]
Unfortunately this didn't fix it for our CommonJS-based package. I'll test out async imports when I have some time to look at this more.

**Comment by joecarl** [comment]
But the OP complains about this error happening when trying to run with tsx. I got exactly the same issue and solved it the way I stated.

Btw @dguenther  I don't understand why lazy loading would work better than normal imports in this case, I'm not saying it won't work, just that I don't understand. Would be interested in an explanation.

**Comment by dguenther** [comment]
Ah, that's fair, I didn't look closely enough at the error message and stack trace, so my issue has a similar error but different cause. When using `require` in a CommonJS package:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in D:\repos\js-datachannel-benchmarks\node_modules\node-datachannel\package.json
    at new NodeError (node:internal/errors:405:5)
    at exportsNotFound (node:internal/modules/esm/resolve:359:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:639:13)
    at resolveExports (node:internal/modules/cjs/loader:567:36)
    at Module._findPath (node:internal/modules/cjs/loader:636:31)
    at Module._resolveFilename (node:internal/modules/cjs/loader:1063:27)
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:121:18)
    at Object.<anonymous> (D:\repos\js-datachannel-benchmarks\src\messages.js:10:25) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}
```

If a package is using `tsc` with `module: node16`, for example (docs here: https://www.typescriptlang.org/docs/handbook/modules/reference.html#interoperability-rules), static imports will be transformed to `require` calls when used in CommonJS modules, so I end up with a similar error as I mentioned above. But as the docs mention, dynamic `import()` calls can be used to import ES modules.

**Comment by unicomp21** [comment]
should I be using tsx w/ a different node version?

**Comment by unicomp21** [comment]
fwiw, I was using node16

**Comment by joecarl** [comment]
Not sure. Anyway node 16 is very outdated today. Try updating to node 20 at least.

**Comment by murat-dogan** [comment]
Hi,
There are multiple questions/problems actually.
First it seems tsx working differently. I changed exports to this;
```
    "exports": {
        ".": "./lib/index.js",
        "./polyfill": "./polyfill/index.js"
    },
```
With these changes it works.

For commonjs modules;
I need to switch to ESM for polyfills. And I dropped support for node versions less than 16.
With current changes, these will work;
```
import nodeDatachannel from 'node-datachannel';
```

and also call with tsx as I wrote above.

For supporting both commonjs and esm, it seems we need some work. But I am not sure if it is worth it.
I found this blog 
https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html

**Comment by dguenther** [comment]
@murat-dogan Thanks for the update! When you have time, would you be able to push a new version with the export change from here? https://github.com/murat-dogan/node-datachannel/commit/92ced431a57162e05a26b0b713177f4afae72051 I have ESM working in my commonJS TS project except for that change.

**Comment by murat-dogan** [comment]
Thanks for the feedback.
I have just released v0.5.1
